File: c/src/output.c
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
#include "output.h"
#include <stdio.h>
#include <stdlib.h>

int output_save_png(const char *filename, int width, int height, const uint8_t *rgba) {
    return stbi_write_png(filename, width, height, 4, rgba, width * 4);
}

int output_assemble_video(const char *pattern, int fps, const char *out_path) {
    char cmd[1024];
    snprintf(cmd, sizeof(cmd),
        "ffmpeg -y -framerate %d -i \"%s\" -c:v libx264 -pix_fmt yuv420p \"%s\"",
        fps, pattern, out_path);
    return system(cmd);
}

File: rust/src/output.rs
use image::RgbaImage;
use std::process::Command;

/// Save an image as a PNG file.
pub fn save_png(filename: &str, img: &RgbaImage) -> Result<(), Box<dyn std::error::Error>> {
    img.save(filename)?;
    Ok(())
}

/// Assemble numbered PNG frames into an MP4 video using ffmpeg.
pub fn assemble_video(pattern: &str, fps: u32, out_path: &str) -> Result<(), Box<dyn std::error::Error>> {
    let status = Command::new("ffmpeg")
        .args([
            "-y",
            "-framerate", &fps.to_string(),
            "-i", pattern,
            "-c:v", "libx264",
            "-pix_fmt", "yuv420p",
            out_path,
        ])
        .status()?;

    if !status.success() {
        return Err(format!("ffmpeg exited with status {}", status).into());
    }
    Ok(())
}
